### PR TITLE
User-Facing Experimental Features

### PR DESCRIPTION
**Pull request by csweichel**
This PR lets users enable "experimental features" within Gitpod.

@gtsiolis proposed that we should have a single "experimental features" flag as compared to surfacing the individual feature flags. This makes perfect sense, as some of those flags interact with each other (you cannot use GP Code without `registry_facade`), and others make no sense to the user directly (e.g. `registry_facade`). Instead, we let the user express the intent that they'd like to use the latest and greatest.

What the latest and greatest is, is something we can then decide. At the moment it's tied to a Gitpod installation, i.e. we can change what "experimental features" are by redeploying.

### How does it work?
When user's enable experimental features, we set a flag in the user's `additionalConfig`. Upon workspace **instance** creation we check if the user has this flag set, and if so, set additional feature flags on the workspace and apply the IDE image of choice.
The concrete feature flags we set when experimental features are enabled are configured on the server in `EXPERIMENTAL_FEATURE_FLAGS`.

Tying this decision to the workspace instance provides the following trade off:
- :+1: we can at any point remove experimental feature flags if we see that they break the user's experience, or worse, impact the overall system negatively.
- :+1: enabling experimental features makes those available also for existing workspaces. Until we make workspace-permanent feature flags (e.g. `full_workspace_backup`) experimental features, this defers the need for marking which workspace supports those new features and which one doesn't. Instead, if the user ticks the "experimental features" box, the system behaves coherently different.

When we have workspace-permanent feature flags (e.g. `full_workspace_backup`) from which there is "no way back", i.e. once a workspace was started (not just created) with that flag, it cannot be started without it anymore, then we need to introduce some more machinery for providing feedback.

This PR also moves the IDE configuration away from the workspace configuration again, meaning you cannot set the IDE in your `.gitpod.yml` "anymore". For now users can choose between Theia and Code, a rather personal preference. Only when we enable custom IDE images, does it make sense to set this on a project level. I'd argue we should not do this before registry facade has graduated to regular production, and isn't a feature flag anymore.

### UX
The UX is far from refined. There's some basic explanation what the experimental features do, but not much. Also, there's no feedback afterwards, i.e. user's can't tell what impact that choice had. We'll probably need some iteration on that.

### How to test?
1. Head over to https://cw-gitpod-enable-feature.staging.gitpod-dev.com/settings/ and enable "experimental features"
2. Start a workspace. `sudo` should work.
3. Change the IDE to `Code` and start a new workspace. It should come up with GP Code instead of Theia.

### CODE REVIEW

**Comment by akosyakov** [comment]
Supporting custom images would be nice to enable some users to use Gitpod Theia from core master.

**Comment by svenefftinge** [comment]
I think it would be great to have a check box per feature. I mean not the internal feature flags, but what the user actually gets. It is more informative and you are able to opt-in to certain things while disabling others (if you don't like them or they break your experience).

Concretely, we should have an "experimental" section with two features now:
 - theia / VS Code
 - root access 

(Not sure about the naming)

**Comment by csweichel** [comment]
> Supporting custom images would be nice to enable some users to use Gitpod Theia from core master.

We could bring it back if you have the right permission. WDYT?

**Comment by csweichel** [comment]
> I think it would be great to have a check box per feature. I mean not the internal feature flags, but what the user actually gets. It is more informative and you are able to opt-in to certain things while disabling others (if you don't like them or they break your experience).
> 
> Concretely, we should have an "experimental" section with two features now:
> 
> * theia / VS Code
> * root access
> 
> (Not sure about the naming)

I agree, individual flags would be more informative. I'd hope though that a better description, possibly a section in the docs would accomplish the same. In there we could explicitly list the features that are experimental right now.

There are a couple of benefits of having a "generic" experimental features flag, rather than splitting them up individually:
- no combinatorial explosion: with just the two options we have right now (Theia, Code, Root Access) that's four combinations; add a single other option (e.g. FWB) and you got eight. This makes the experience we collect "in the wild" less representative (because users might choose combinations we don't really want going forward), and harder to support.
- opt-in once, add many more: if users express their intent to explore experimental features, we can add new features over time without users having to opt in again. Unlike the last time we tried registry-facade, we'd now have a group of users who have explicitly expressed their consent/interest in trying this out. If, however, they've just ticked "Root in Workspaces" instead of "experimental features" we don't get that expression of intent.

**Comment by svenefftinge** [comment]
Ok, understood. 
Experimental features that need settings (e.g. Theia/Code/Your IDE) should then go below the section and only be enabled when the major "experimental features" toggle is enabled. We can have a link to the documentation section describing all the things that you enable by enabling this.

**Comment by csweichel** [comment]
We'll rename it to "feature preview"

**Comment by akosyakov** [comment]
> We could bring it back if you have the right permission. WDYT?

I'm fine with any way. It is an experimental feature, there is no guarantee that we won't remove custom images later.

**Comment by csweichel** [comment]
I've renamed it to "feature preview"
image

and brought the "image" field back, if the user has the "ide-settings" permission.
image

**Comment by jankeromnes** [comment]
Thought: "Preview" or "Feature Preview" can be confusing because there is already a feature called "Preview" (in the IDE), and just the name "Feature Preview" does not really tell me what to expect.

Maybe we can just call it what it is, i.e. "Experimental Features" and "[ ] Enable Experimental Features"?

See also https://trends.google.com/trends/explore?q=experimental%20features,feature%20preview

**Comment by csweichel** [comment]
> Thought: "Preview" or "Feature Preview" can be confusing because there is already a feature called "Preview" (in the IDE), and just the name "Feature Preview" does not really tell me what to expect.
> 
> Maybe we can just call it what it is, i.e. "Experimental Features" and "[ ] Enable Experimental Features"?
> 
> See also https://trends.google.com/trends/explore?q=experimental%20features,feature%20preview

We just renamed it from "experimental features" to feature preview. I'd say we stick with that name for now.

**Comment by csweichel** [comment]
> Thought: "Preview" or "Feature Preview" can be confusing because there is already a feature called "Preview" (in the IDE), and just the name "Feature Preview" does not really tell me what to expect.
> 
> Maybe we can just call it what it is, i.e. "Experimental Features" and "[ ] Enable Experimental Features"?
> 
> See also https://trends.google.com/trends/explore?q=experimental%20features,feature%20preview

We choose the term "feature preview" over "experimental features" because we hope it will encourage more users to actually activate them. Also, what we're offering there is more mature than a mere experiment: it's the next generation of product features, which just aren't quite ready yet for mass consumption for one reason or another.
This reasoning does not preclude that we also enable actually experimental features at times (see discussion why we should have a single flag rather than letting users choose one for each feature).

**Comment by jankeromnes** [comment]
Thanks for explaining the rationale.

Then maybe "Beta Features" would sound more production-ready and enticing, yet a bit more honest & transparent than "Feature Preview"?

**Review by gtsiolis — COMMENTED** [review]
Looks good! Added a minor suggestion.

**Comment by gtsiolis** [comment]
Thanks @Chris for breaking this down! 🙏 

> Supporting custom images ...

@akosyakov I haven't used this, it looks like an interesting feature but I'm not sure if this should be exposed to every user instead of the admin. Has this been discussed before? **Question**: Do we expect users or consider safe to inject any container image for the IDE while using Gitpod.io?

> I think it would be great to have a check box per feature. ...

@svenefftinge it certainly can be valuable to have multiple toggles for each feature, especially for user-facing features, but this wouldn't be practical or useful for more than a couple of features at a time. To add some more context from the discussion with @csweichel, the intent is to allow early adopters or advanced users to try out upcoming features before their public release so that we can catch bugs earlier and collect feedback as early as possible.

FWIW, Enabling _all_ feature previews with a single toggle control can expose more users to a feature that otherwise could have skipped or avoided for several reasons thus leading to decreased usage and exposure time of a feature.

💡  **Idea**: We could introduce a way of collecting feedback within the product, regardless this PR, or point the users to the issue tracker for suggestions and feedback after enabling _Feature Preview_. WDYT?

> We'll rename it to ...

@csweichel +1 for renaming this to _Feature Preview_ as this is also a common name used in other services and will help [minimize cognitive load](https://www.nngroup.com/articles/minimize-cognitive-load/) by relying on existing mental models for users.

FWIW, I think it doesn't matter much how we name this here as we can always improve it later. What matters is that we can commonly refer to this feature with the same name.

I like the [break down](https://docs.microsoft.com/en-us/powerapps/maker/canvas-apps/working-with-experimental-preview) of the difference between experimental and preview features in Power Apps:

| Stage | When should I use it? | Can I use it with confidence? Is it enabled by default for new apps? |
|-|-|-|
| Experimental | If you're an early adopter, see something useful to you, and would like to help test the feature. | No. Experimental features can radically change or completely disappear at any time. For this reason the feature is not enabled by default and you must explicitly opt in to use it. |
| Preview | New apps automatically include this feature but it can still be turned off. Start enabling and testing in existing apps because this feature will be eventually turned on for them too. | Yes. This feature is on track to become a permanent part of the product. You may want to turn it off if you run into a problem. Please report issues; this is the main reason the feature is in Preview. |

❓  **Question:** The features discussed here seem like _Preview Features_ that will end up in the product in the end, right? I could be wrong but we are not trying them out to see how they work, right?

Last but not least, regarding the UX when enabling Feature Preview, to avoid abrupt changes in the user interface, like the IDE selection below, we could introduce a) a toggle button instead of a checkbox and b) move the extra settings in a sidebar to group and expand settings functionalities like IDE selection. However, this is probably out of the scope here, I'm just adding this for future reference and further discussion.

| BEFORE | AFTER |
|-|-|
| image | image |

Hopefully, some of the points mentioned here are helpful and didn't cause more confusion. 😅

File: components/dashboard/src/components/feature-settings.tsx [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    componentDidUpdate(prevProps: FeatureSettingsProps) {
+        if (this.props.user !== prevProps.user) {
+            this.setStateFromUser((this.props.user));
+        }
+    }
+
+    render() {
+        return <Grid xs={12}>
+            <div><small>Enabling the feature preview gives you access to Gitpod's latest functionality and features. Some of those you'll notice immediately, others operate behind the scenes. <i>Here be dragons.</i></small></div>
```
[the remark this answers is not in the document]
Comment (csweichel) [reply components/dashboard/src/components/feature-settings.tsx]: I've updated the text. Actionable links are a much more sensible choice compared to some 18th century cartographer reference :D

**Comment by csweichel** [comment]
I've updated the description with @gtsiolis proposal:
image

**Review by svenefftinge — APPROVED** [review]
LGTM

**Pull request merged**